---
title: "Project Network-based Data Analysis"
author: "Annalisa Xamin"
output: 
  pdf_document:
    latex_engine: xelatex
    toc: true
  html_document:
    toc: true
    df_print: paged
always_allow_html: true
---
```{r, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
# knitr::opts_chunk$set(cache = TRUE)
setwd("~/Repo_Git/NBDA/Code")
```


```{r packages, warning=FALSE, results='hide', message=FALSE, include=FALSE}
# Function to install and load packages
install_and_load <- function(pkg, bioc = FALSE) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    if (bioc) { # Bioconductor
      if (!require("BiocManager", quietly = TRUE)) {
        install.packages("BiocManager")
        library(BiocManager)
      }
      BiocManager::install(pkg, ask = FALSE)
    } else { # CRAN 
      install.packages(pkg)
    }
    library(pkg, character.only = TRUE)
  } else {
    library(pkg, character.only = TRUE)
  }
}

# Bioconductor packages
bioc_packages <- c("GEOquery", "rScudo", "limma", "EnhancedVolcano",
                   "pROC", "hgu133a.db"
                   )

# CRAN packages
cran_packages <- c(
  "tidyverse", "biomaRt", "dplyr", "plotly", "sessioninfo",
  "ggplot2", "factoextra", "useful", "umap", "ggpubr",
  "magrittr", "cluster", "randomForest", "RColorBrewer", "caret",
  "pheatmap", "MASS"
)

# Install/load all packages
lapply(bioc_packages, install_and_load, bioc = TRUE)
lapply(cran_packages, install_and_load)

# Colors palette 
my_colors <- c("#2e005d", "#5c008b", "#8e008b", "#ff8300", "#ff6200", "#d1105a", "#05a8aa")
```

```{r session info, warning=FALSE, include=FALSE}
print('Reproducibility information:') 
Sys.time()
proc.time()
options(width = 120)
session_info()
```

# Data selection

The analysis will use the dataset [GSE20437](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20437) obtained from GEO.The dataset is generated from Affymetrix HU133A microarrays and contains 42 tissue samples. 

In detail, the data includes:

- 18 reduction mammoplasty (RM) breast epithelium samples, 

- 18 histologically normal (HN) epithelial samples from breast cancer patients (9 ER+ and 9 ER-), and

- 6 histologically normal epithelial samples from prophylactic mastectomy patients. 

Note that sample numbers correspond to individual patient samples.

```{r retrieve data from GEO}
# download the GSE20437 expression data series
#gse <- getGEO("GSE20437", destdir= './data/', getGPL = F)

# load the local copy of the data
gse <- getGEO(file = "./data/GSE20437_series_matrix.txt.gz", getGPL = FALSE)
```

```{r}
# getGEO returns a list of expression objects, but...
length(gse) 
# shows us there is only one object in it. 
# We assign it to the same variable.
#gse <- gse[[1]] # run only if you download data and 
# if you don't use the local copy
```

```{r retrieve metadata}
# extract metadata
metadata <- data.frame(gse@phenoData@data)
```

```{r save metadata to csv file, include=FALSE}
# save metadata to csv file
write.csv(metadata, file = 'output/metadata.csv', row.names = TRUE)
```

```{r}
expr(gse[1])
```

## Annotation of probesets from a GEO dataset to gene symbols
For the later analysis, it is useful to annotate the probe sets of the GEO data set to gene symbol. 
To do that, first we extract the name of the probes and then, we perform the annotation using biomaRt.

```{r}
id_to_annotate <- rownames(gse@assayData[["exprs"]])
# extract id to annotate

# annotation
mart <- useMart("ENSEMBL_MART_ENSEMBL")
mart <- useDataset("hsapiens_gene_ensembl", mart)
annotLookup <- getBM(
  mart = mart,
  attributes = c(
    "affy_hg_u133_plus_2",
    "ensembl_gene_id",
    "gene_biotype",
    "external_gene_name",
    "description"),
  filter = "affy_hg_u133_plus_2",
  values = id_to_annotate,
  uniqueRows=TRUE)
```

```{r}
head(annotLookup)
```

```{r}
indicesLookup <- match(rownames(gse@assayData[["exprs"]]), annotLookup$affy_hg_u133_plus_2)
#head(annotLookup[indicesLookup, "external_gene_name"])
```
```{r}
dftmp <- data.frame(rownames(gse@assayData[["exprs"]]), 
                    annotLookup[indicesLookup, 
                                c("affy_hg_u133_plus_2", "external_gene_name")])
head(dftmp, 20)
```

```{r}
length(rownames(gse@assayData[["exprs"]]))
```

```{r}
table(dftmp[,1] == dftmp[,2])
```

## Explore the groups in the dataset
```{r plot groups, warning=FALSE}
p <- ggplot(metadata, aes(x=disease.state.ch1, fill=specimen.ch1))+
  geom_bar()+
  scale_fill_manual(values = my_colors[c(1,4,6,7)])
p + labs(x = "Group") 

```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/01_plot_groups.pdf")
p <- ggplot(metadata, aes(x=disease.state.ch1, fill=specimen.ch1))+
  geom_bar()+
  scale_fill_manual(values = my_colors[c(1,4,6,7)])
p + labs(x = "Group") 
dev.off()
```


```{r, include=FALSE}
head(metadata)
```


# Exploratory analysis
```{r}
# show what we have:
show(gse)
```

The actual expression data are accessible in the `exprs` section of `gse`, an Expression Set and the generic data class that BioConductor uses for expression data.

```{r inspect exprs}
head(exprs(gse)) 
```

```{r length exprs, include = FALSE}
length(exprs(gse))
```

To conveniently access the data rows and columns present in `exprs(gse)`, this matrix is assigned to its own variable `ex`.

```{r}
# exprs (gse) is a matrix that we can assign to its own variable, to
# conveniently access the data rows and columns
ex <- exprs(gse)
dim(ex) # 42 sample, 22283 genes
```

The dataset contains gene expression data of 22283 genes (rows) from 42 patients (columns).

```{r, include=FALSE}
colnames(ex)
```

```{r, include = FALSE}
rownames(ex)
```

```{r, include = FALSE}
ex[1:5,]
```

```{r, include=FALSE}
metadata$description
# contains case number
```


## Pre-processing

```{r boxplot original data, warning=FALSE}
# Analyze value distributions
boxplot(ex, main = 'Boxplot of the data before normalization',
        xlab = "Samples",
        ylab = "Expression Value",
        varwidth = TRUE
        )
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/02_boxplot_no_norm.pdf")
boxplot(ex, main = 'Boxplot of the data before normalization',
        xlab = "Samples",
        ylab = "Expression Value",
        varwidth = TRUE
        )
dev.off()
```

The boxplot shows that scaling is necessary. So, in this case, I try to apply a log transformation to the data.

```{r boxplot log transformation, fig.cap="Boxplot of the data after applying a logarithmic transformation"}
ex2<-log(ex)
ex2 <- na.omit(as.matrix(ex2))
#dim(ex2) # 22283    42 same as before
boxplot(ex2, main = 'Boxplot of the data after applying a logarithmic transformation',
        xlab = "Samples",
        ylab = "Expression Value"
        )
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/03_boxplot_log_no_norm.pdf")
boxplot(ex2, main = 'Boxplot of the data after applying a logarithmic transformation',
        xlab = "Samples",
        ylab = "Expression Value"
        )
dev.off()
```

From the boxplot after the log transformation, I can see that there is some variation in the median of the samples. So, one of the simplest normalization strategies is to align the log values so that all channels have the same median. A convenient choice is zero so that positive or negative values reflect signals above or below the median for a particular channel.

```{r}
normalized.log.ex=scale(log(ex))

# boxplot post median normalization on ex
boxplot(normalized.log.ex, 
        main = 'Boxplot of the data after median normalization',
        xlab = "Samples",
        ylab = "Expression Value")
```

## PCA
PCA is a dimensionality reduction technique that allows to condense thousands of dimensions into just two or three.For the dataset's samples, the PCA scores display the coordinates in relation to these additional dimensions. 

```{r pca}
pca <- prcomp(t(normalized.log.ex))

summary(pca)
screeplot(pca)
```

To get the summary of the PCA and the plot showing the variance explained by the first 10 components, it is possible to use the functions commented in the chunks above.

However, using `ggplot2` and `factoextra` packages is possible to get a more concise and informative plot reporting the same information.

```{r variance explained by pca}
pcaVar <- get_eig(pca)
pcaVar <- pcaVar$variance.percent[1:10]
screeDf <- data.frame("Dimensions" = as.factor(seq(1,10)),
                      "Percentages" = pcaVar,
                      "Labels" = paste(round(pcaVar, 2), "%"))

p <- ggplot(data = screeDf, aes(x=Dimensions, y=Percentages))+
  geom_bar(stat = "identity", fill = "#d1105a")+
  geom_text(aes(label=Labels), vjust=-0.5, color="black", size=3.6)+
  ggtitle("Scree Plot")+
  ylab("Percentage of variance explained")+
  scale_x_discrete(labels = as.factor(seq(1,10)))
p
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/05_ScreePlot.pdf")
p
dev.off()
```

The scree plot shows that the first dimensions on the left are the more important because the percentage of variance explained by them is higher. The remaining principal components account for a very small proportion of the variability and are probably unimportant.

Let's try to plot the PCA, looking if we can see a separation between Control and Breast Cancer groups.

```{r PCA for components 1 and 2 control VS Breast cancer without labels}
# draw PCA plot control VS breast cancer
group <- c(rep("cadetblue1",18), rep("red",18), rep("cadetblue1",6) ) 
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c("cadetblue1","red"), legend = c("Control", "Breast Cancer"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/06_PCA_tumorVScontrol.pdf")
# draw PCA plot control VS breast cancer
group <- c(rep("cadetblue1",18), rep("red",18), rep("cadetblue1",6) ) 
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c("cadetblue1","red"), legend = c("Control", "Breast Cancer"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

```{r PCA for components 1 and 2 control VS Breast cancer with labels, include=FALSE}
# draw PCA plot control VS breast cancer
group <- c(rep("cadetblue1",18), rep("red",18), rep("cadetblue1",6) ) 
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5) # add labels to points
legend("topleft", col=c("cadetblue1","red"), legend = c("Control", "Breast Cancer"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/07_PCA_tumorVScontrol_labels.pdf")
# draw PCA plot control VS breast cancer
group <- c(rep("cadetblue1",18), rep("red",18), rep("cadetblue1",6) ) 
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5) # add labels to points
legend("topleft", col=c("cadetblue1","red"), legend = c("Control", "Breast Cancer"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

Let's try to add the control subtypes. The vector group used in the PCA plot is based on the data. The samples corresponding to the colors are the following:

-   **Light blue**: reduction mammoplasty (RM) breast epithelium samples

-   **Red**: histologically normal (HN) epithelial samples from breast cancer patient

-   **Purple**: histologically normal breast epithelium (NlEpi) from prophylactic mastectomy patient samples

```{r PCA for components 1 and 2 control subtypes without labels}
# draw PCA plot
group <- c(rep("cadetblue1",18), rep("red",18), rep("purple",6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c("cadetblue1","red","purple"), legend = c("Reduction Mammoplasty", "Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/08_PCA_subtypes_control.pdf")
group <- c(rep("cadetblue1",18), rep("red",18), rep("purple",6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c("cadetblue1","red","purple"), legend = c("Reduction Mammoplasty", "Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

```{r PCA for components 1 and 2 control subtypes with labels, include=FALSE}
# draw PCA plot
group <- c(rep("cadetblue1",18), rep("red",18), rep("purple",6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5) # add labels
legend("topleft", col=c("cadetblue1","red","purple"), legend = c("Reduction Mammoplasty", "Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/09_PCA_subtypes_control_labels.pdf")
# draw PCA plot
group <- c(rep("cadetblue1",18), rep("red",18), rep("purple",6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5) # add labels
legend("topleft", col=c("cadetblue1","red","purple"), legend = c("Reduction Mammoplasty", "Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

Then, I try to see if there is a separation also inside different types of Breast Cancer.
```{r pca with all subtypes without labels}
# draw PCA plot with all subtypes
group <- c(rep(my_colors[7],18), rep(my_colors[4],9), rep(my_colors[1],9), rep(my_colors[6],6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c(my_colors[7],my_colors[4],my_colors[1],my_colors[6]), legend = c("Reduction Mammoplasty", "ER+ Breast Cancer", "ER- Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/10_PCA_all_subtypes.pdf")
# draw PCA plot with all subtypes
group <- c(rep(my_colors[7],18), rep(my_colors[4],9), rep(my_colors[1],9), rep(my_colors[6],6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$data), cex=0.75)
legend("topleft", col=c(my_colors[7],my_colors[4],my_colors[1],my_colors[6]), legend = c("Reduction Mammoplasty", "ER+ Breast Cancer", "ER- Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

```{r pca with all subtypes with labels, include=FALSE}
# draw PCA plot with all subtypes
group <- c(rep(my_colors[7],18), rep(my_colors[4],9), rep(my_colors[1],9), rep(my_colors[6],6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5)
legend("topleft", col=c(my_colors[7],my_colors[4],my_colors[1],my_colors[6]), legend = c("Reduction Mammoplasty", "ER+ Breast Cancer", "ER- Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/11_PCA_all_subtypes_labels.pdf")
# draw PCA plot with all subtypes
group <- c(rep(my_colors[7],18), rep(my_colors[4],9), rep(my_colors[1],9), rep(my_colors[6],6) ) # vector of colors based on the order of my data
plot(pca$x[,1], pca$x[,2], xlab="PCA1", ylab="PCA2", main="PCA for components 1 and 2", type="p", pch=10, col=group)
text(pca$x[,1], pca$x[,2], rownames(pca$x), cex=0.5)
legend("topleft", col=c(my_colors[7],my_colors[4],my_colors[1],my_colors[6]), legend = c("Reduction Mammoplasty", "ER+ Breast Cancer", "ER- Breast Cancer", "Prophylactic Mastectomy"),
    pch = 20, bty='n', cex=.55)
dev.off()
```

### Interactive PCA plot
Let's try to explore an interactive PCA plot.

```{r fig1 interactive PCA plot, message=FALSE}
components<-pca[["x"]]
components<-data.frame(components)
type<-c(rep("RM", 18), rep("HN",18), rep("NlEpi",6))
components<-cbind(components, type )

fig <- plot_ly(components, x=~PC1, y=~PC2, 
               color=type,colors=c('cadetblue1', 'red','purple'), 
               type='scatter',mode='markers')
fig
```

```{r fig2 interactive PCA plot, message=FALSE}
fig2 <- plot_ly(components, x=~PC1, y=~PC2, z=~PC3, 
                color=type, colors=c('cadetblue1', 'red','purple'),
                mode='markers', marker = list(size = 4))
fig2
```


```{r fig3 interactive PCA plot, message=FALSE}
fig3 <- plot_ly(components, x=~PC1, y=~PC3, 
                color=type, colors=c('cadetblue1', 'red','purple'),
                type='scatter',mode='markers')
fig3
```

```{r}
set.seed(123)  
umap_result <- umap(t(normalized.log.ex))

umap_df <- as.data.frame(umap_result$layout)
colnames(umap_df) <- c("UMAP1", "UMAP2")

# Add sample types to the UMAP data frame
umap_df$type <- c(rep("RM", 18), rep("HN",18), rep("NlEpi",6))

ggplot(umap_df, aes(x = UMAP1, y = UMAP2, color = type)) +
  geom_point(size = 3, alpha = 0.8) +
  scale_color_manual(values = c("RM" = "cadetblue1", 
                                "HN" = "red", 
                                "NlEpi" = "purple")) +
  theme_minimal() +
  labs(title = "2D UMAP Projection of GSE20437", x = "UMAP1", y = "UMAP2")

```

```{r, include=FALSE}
pdf("plots/12_UMAP_plot.pdf")
ggplot(umap_df, aes(x = UMAP1, y = UMAP2, color = type)) +
  geom_point(size = 3, alpha = 0.8) +
  scale_color_manual(values = c("RM" = "cadetblue1", 
                                "HN" = "red", 
                                "NlEpi" = "purple")) +
  theme_minimal() +
  labs(title = "UMAP Projection of GSE20437", x = "UMAP1", y = "UMAP2")
dev.off()

```

# Clustering
## K-means
```{r K-means}
set.seed(1)
k <- 2 # number of clusters

kmeans_result <- kmeans(t(normalized.log.ex),k)
table(kmeans_result$cluster) # tells how many samples were assigned to each cluster
```

```{r plot kmeans, include=FALSE}
# save plot as pdf
pdf("plots/12_kmeans_2_labels.pdf")
plot(kmeans_result, data=t(normalized.log.ex)) + geom_text(aes(label=colnames(ex)),hjust=0,vjust=0)
```

```{r plot kmeans with metadata title}
plot(kmeans_result, data=t(normalized.log.ex)) + geom_text(aes(label=metadata$disease.state.ch1),hjust=0,vjust=0)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/13_kmeans_2_metadata_title.pdf")
plot(kmeans_result, data=t(normalized.log.ex)) + geom_text(aes(label=metadata$disease.state.ch1),hjust=0,vjust=0)
dev.off()
```

```{r cluster plot kmeans 2 clusters}
fviz_cluster(kmeans_result, data = t(normalized.log.ex),
              palette = c("#FF6666", "#33cccc"), 
             geom = "point",
             ellipse.type = "convex", 
             ggtheme = theme_bw()
             )
```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/14_kmeans_2_cluster_plot.pdf")
fviz_cluster(kmeans_result, data = t(normalized.log.ex),
              palette = c("#FF6666", "#33cccc"), 
             geom = "point",
             ellipse.type = "convex", 
             ggtheme = theme_bw()
             )
dev.off()
```

Let's try increasing the number of clusters.
```{r}
set.seed(1)
k <- 4 # number of clusters

kmeans_result <- kmeans(t(normalized.log.ex),k)
table(kmeans_result$cluster) # tells how many samples were assigned to each cluster
```

```{r plot kmeans with metadata specimen}
plot(kmeans_result, data=t(normalized.log.ex)) + geom_text(aes(label=metadata$specimen.ch1),hjust=0,vjust=0)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/15_kmeans_4_metadata_title.pdf")
plot(kmeans_result, data=t(normalized.log.ex)) + geom_text(aes(label=metadata$specimen.ch1),hjust=0,vjust=0)
dev.off()
```

```{r cluster plot kmeans 4 clusters}
fviz_cluster(kmeans_result, data = t(normalized.log.ex),
             palette = c("#FF6666", "#99C666", "#33cccc", "#cc66ff"), 
             geom = "point",
             ellipse.type = "convex", 
             ggtheme = theme_bw()
             )
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/16_kmeans_4_cluster_plot.pdf")
fviz_cluster(kmeans_result, data = t(normalized.log.ex),
             palette = c("#FF6666", "#99C666", "#33cccc", "#cc66ff"), 
             geom = "point",
             ellipse.type = "convex", 
             ggtheme = theme_bw()
             )
dev.off()
```

## Hierarchical


```{r determine optimal number of cluster}
# Elbow method
fviz_nbclust(t(normalized.log.ex), FUN = hcut, method = "wss") 
## seems setting number of clusters equal to 2
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/17_elbow_plot.pdf")
fviz_nbclust(t(normalized.log.ex), FUN = hcut, method = "wss") 
dev.off()
```

```{r}
# Silhouette method
fviz_nbclust(t(normalized.log.ex), FUN = hcut, method = "silhouette")
## seems setting number of clusters equal to 2
```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/18_Silhouette_plot.pdf")
fviz_nbclust(t(normalized.log.ex), FUN = hcut, method = "silhouette")
dev.off()
```

We use the Gap statistic to calculate the goodness of clustering.
```{r}
# Gap Statistic Method
gap_stat <- clusGap(t(normalized.log.ex), FUN = hcut, nstart = 25, K.max = 10, B = 50)
# K.max -> the maximum number of clusters to consider
# B -> number of Monte Carlo samples

fviz_gap_stat(gap_stat)
```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/19_Gap_plot.pdf")
fviz_gap_stat(gap_stat)
dev.off()
```


```{r}
hc_result <- dist(t(normalized.log.ex)) %>% hclust(method = "ave")
hc_result2<- dist(t(normalized.log.ex), method="euclidean") %>% hclust( method = "complete") 
hc_result3 <- dist(t(normalized.log.ex)) %>% hclust(method = 'single')

k_hc <- 2 # optimal number of clusters

groups <- cutree(hc_result, k=k_hc)
table(groups,type)

groups2<-cutree(hc_result2, k=k_hc)
table(groups2,type)

groups3 <- cutree(hc_result3,k=k_hc)
table(groups3,type)
```
```{r}
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (average)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
```

```{r}
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result2, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (complete)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
```

```{r}
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result3, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (single)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/20_Hierarchical_ave_plot.pdf")
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (average)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
dev.off()
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/21_Hierarchical_complete_plot.pdf")
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result2, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (complete)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
dev.off()
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/22_Hierarchical_single_plot.pdf")
# Set up layout with extra space below the plot
par(mar = c(6, 4, 4, 2))   # increase bottom margin
par(xpd = TRUE)            # allow text outside plot region

# Plot dendrogram
plot(hc_result3, hang = -1, labels = type, 
     main = 'Hierarchical clustering dendrogram (single)')
rect.hclust(hc_result, k = 2, border = 2) # red boxes to show groups

# Add custom legend box below the plot
text(x = mean(par("usr")[1:2]), y = par("usr")[3] - 10,
     labels = paste(
       "RM = Reduction Mammoplasty breast epithelium\n",
       "HN = Histologically normal epithelial samples from breast cancer patients (ER+ and ER-)\n",
       "NlEpi = Normal epithelial samples from prophylactic mastectomy patients"
     ),
     cex = 0.8, adj = 0.5)
dev.off()
```

## Random forest

```{r plot random forest error rate vs number of trees}
set.seed(1234)
rf <- randomForest(x=t(normalized.log.ex), y=as.factor(type), ntree=1000)
plot(rf, main = "Random Forest")
```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/23_RF_Error_rate_VS_n_trees.pdf")
plot(rf, main = "Random Forest")
dev.off()
```

The plot above shows how the error rate changes according to the number of trees. This allows to track how model accuracy improves (or doesn't) as more trees are added.
The black line represent the overall OOB (out-of-bag) error rate, while the other colored lines report the class specific error rates.

```{r, include=FALSE}
tail(plot(rf))
```


```{r}
# a trivial test
predict(rf, t(normalized.log.ex[, 1:5]))
```

```{r top 30 features random forest}
# Get importance values
rf_importance <- importance(rf)
# Extract probe IDs in order of importance
ordered_probes <- rownames(rf_importance)[order(rf_importance[, 1], decreasing = TRUE)]
# Ensure correct mapping exists
probe_to_symbol <- annotLookup[, c("affy_hg_u133_plus_2", "external_gene_name")]
colnames(probe_to_symbol) <- c("probe", "symbol")
probe_symbols <- setNames(probe_to_symbol$symbol, probe_to_symbol$probe)

top_vars <- head(ordered_probes, 30)  # Show top 30 features
imp_df <- data.frame(
  probe = top_vars,
  importance = rf_importance[top_vars, 1],
  gene = ifelse(top_vars %in% names(probe_symbols), probe_symbols[top_vars], top_vars)
)

ggplot(imp_df, aes(x = reorder(gene, importance), y = importance)) +
  geom_bar(stat = "identity", fill = "#2e005d") +
  coord_flip() +
  labs(title = "Top 30 Variable Importance (Random Forest)", x = "Gene", y = "Importance") +
  theme_minimal(base_size = 12)
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/24_RF_Variable_importance_top_30.pdf")
ggplot(imp_df, aes(x = reorder(gene, importance), y = importance)) +
  geom_bar(stat = "identity", fill = "#2e005d") +
  coord_flip() +
  labs(title = "Top 30 Variable Importance (Random Forest)", x = "Gene", y = "Importance") +
  theme_minimal(base_size = 12)
dev.off()
```

```{r, include=FALSE}
imp_df
```

```{r save variable importance rf, include=FALSE}
write.csv(imp_df, file = "output/rf_top_variable_importance.csv", row.names = FALSE)
```


## Heatmap

```{r heatmap}
# Select top N important probes
top_n <- 30  
top_probes <- imp_df$probe[1:top_n]

heatmap_data <- normalized.log.ex[top_probes, ]
# Add gene symbols
gene_symbols <- imp_df$gene[1:top_n]
rownames(heatmap_data) <- gene_symbols

pheatmap(heatmap_data,
         scale = "row",  # normalize expression within each gene
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 8,
         main = "Top Random Forest Genes Heatmap")
```

```{r, include=FALSE}
# save plot as pdf
pdf("plots/25_Heatmap_top_30.pdf")
pheatmap(heatmap_data,
         scale = "row",  # normalize expression within each gene
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 8,
         main = "Top Random Forest Genes Heatmap")
dev.off()
```

```{r heatmap with specimen}
# Ensure sample labels are aligned with expression data
sample_types <- metadata$type
names(sample_types) <- colnames(normalized.log.ex)

# Subset for current samples
sample_types <- sample_types[colnames(heatmap_data)]

# Set rownames to match sample IDs
rownames(metadata) <- metadata$geo_accession
# Create annotation dataframe
annotation_col <- data.frame(Type = metadata[colnames(heatmap_data), "specimen.ch1"])
rownames(annotation_col) <- colnames(heatmap_data)

# Convert to factor
annotation_col$Type <- factor(annotation_col$Type)
type_levels <- levels(annotation_col$Type)
palette_colors <- brewer.pal(n = length(type_levels), name = "Set2")  # or "Dark2", "Paired", etc.
names(palette_colors) <- type_levels

ann_colors <- list(Type = palette_colors)

pheatmap(heatmap_data,
         scale = "row",
         annotation_col = annotation_col,
         annotation_colors = ann_colors,
         show_colnames = TRUE,
         main = "Heatmap Top Random Forest Genes by Specimen")
```
```{r, include=FALSE}
# save plot as pdf
pdf("plots/26_Heatmap_top_30_with_specimen.pdf")
pheatmap(heatmap_data,
         scale = "row",
         annotation_col = annotation_col,
         annotation_colors = ann_colors,
         show_colnames = TRUE,
         main = "Heatmap Top Random Forest Genes by Specimen")
dev.off()
```

## Feature selection

### LDA

```{r}
# samples as rows, genes as columns
lda_data <- t(heatmap_data)
colnames(lda_data) <- make.names(colnames(lda_data), unique = TRUE)

# Use disease state or source name as grouping
group <- metadata[colnames(heatmap_data), "specimen.ch1"]
group <- factor(group)

lda_model <- lda(group ~ ., data = as.data.frame(lda_data))
# Inspect coefficients and proportions
lda_model
```

```{r}
lda_pred <- predict(lda_model)

# Get coordinates of samples in LDA space
lda_df <- data.frame(lda_pred$x)
lda_df$Group <- group

ggplot(lda_df, aes(x = LD1, y = LD2, color = Group)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "LDA of Top RF Genes", x = "LD1", y = "LD2")
```
```{r}
table(Predicted = lda_pred$class, Actual = group)
mean(lda_pred$class == group)  # accuracy
```


```{r}
# samples as rows, genes as columns
lda_data <- t(heatmap_data)
colnames(lda_data) <- make.names(colnames(lda_data), unique = TRUE)

# Use disease state or source name as grouping
group_disease_state <- metadata[colnames(heatmap_data), "disease.state.ch1"]
group_disease_state <- factor(group_disease_state)

lda_model_disease_state <- lda(group ~ ., data = as.data.frame(lda_data))
# Inspect coefficients and proportions
lda_model_disease_state

levels(group_disease_state)

lda_pred_disease_state <- predict(lda_model_disease_state)


# Get the predicted LD1 values
scores <- lda_pred_disease_state$x[, 1]  # LD1
true_labels <- group_disease_state       # true classes

# Generate ROC curve
roc_obj <- roc(response = true_labels, predictor = scores)

# Plot it
plot(roc_obj, col = "#2E9FDF", lwd = 3, main = "ROC Curve for LDA")
auc_value <- auc(roc_obj)
legend("bottomright", legend = paste("AUC =", round(auc_value, 3)), col = "#2E9FDF", lwd = 3)
```



### LASSO

### SCUDO

## Functional enrichment analysis

## Network-based Analysis


